### PR TITLE
C++ 11 for (auto& elem : range)

### PR DESCRIPTION
**Pull request by eisber**
this also includes the JSON PR

### CODE REVIEW

File: cs/cs/Serializer/VowpalWabbitJsonSerializer.cs [hunk 1]
```diff
@@ -0,0 +1,304 @@
...
+using VW.Serializer.Intermediate;
+
+namespace VW.Serializer
+{
+    /// <summary>
+    /// A deserializer from JSON to Vowpal Wabbit native examples.
+    /// </summary>
+    public sealed class VowpalWabbitJsonSerializer
+    {
+        internal const string FeatureIgnorePrefix = "_";
```
Comment (lhoang29) [inline cs/cs/Serializer/VowpalWabbitJsonSerializer.cs]: why not put this in VowpalWabbitConstants and perhaps under a namespace/class called Serialization? Then you have all of these constant strings defined in the same place.

File: vowpalwabbit/interactions.h [hunk 2]
```diff
@@ -128,7 +124,7 @@ template<class R>
   const uint64_t  weight_mask   = all.reg.weight_mask;
   weight* weight_vector = all.reg.weight_vector;
 
-  inner_data<R> id ={dat, offset, weight_mask, weight_vector};
+  // inner_data<R> id ={dat, offset, weight_mask, weight_vector};
```
Comment (lhoang29) [inline vowpalwabbit/interactions.h]: nit: remove commented code

File: vowpalwabbit/feature_group.h [hunk 3]
```diff
@@ -151,13 +322,15 @@ struct features { // the core definition of a set of features.
     slice.delete_v();
     return true;
   }
+
+  features& operator=(const features& rhs)
```
Comment (lhoang29) [inline vowpalwabbit/feature_group.h]: nice change

**Comment by lhoang29** [comment]
Looks like I've already reviewed most of the C# side previously and the rest involves mainly C++ 11 refactoring. My main comment (as expected) is the pervasive use of 'auto'. IMHO using explicit types where needed makes code inspection and readability much easier because:

1. Via non-IDE text editors or browsers, locating code without knowing the type can be very difficult.
2. Although some IDEs like Visual Studio allow you to navigate to type definition on "auto" or "var" keywords, it's possible that other packages may not have the same support (for C# it's not a big issue because almost everyone uses VS for development).
3. Even in the cases where we don't care what the type is (i.e. as long as it implements some interfaces) which seems to be the main argument for 'auto', for code debugging we'd still need to find the actual resolved type. Without a step-through debugger this would again be difficult to do.

For an open-source project with many contributors where the code is rapidly changing, I'd vote for using auto only when necessary (e.g. lambda types) or if the type is clear from a glance (e.g. auto b = new Foo()) .

**Comment by JohnLangford** [comment]
How simple can the full type be made?

-John

**Comment by lhoang29** [comment]
You'd just change 'auto' to the actual type of the object/variable. The rest can stay the same (including the c++ 11 foreach syntax). e.g: `for (int& val : val_vector) ...`

**Comment by eisber** [comment]
I think there are 2 more reasons to use auto:


-          If the types are awkward to write and not sure if anybody actually cares (e.g. vector<string>::iterator)

-          Eases refactoring. (e.g. if one wants to change the type of feature indices)

Markus

**Comment by JohnLangford** [comment]
Hence my question: How simple can the full type be?

-John

**Comment by lhoang29** [comment]
Hmm, I tend to agree with the first one for well known built-in types. The second seems like necessary plumbing work, plus the needs for refactoring is probably less than that for inspection & debugging.

**Comment by lhoang29** [comment]
If the full type is too long but non-trivial, we can also use typedef

**Comment by eisber** [comment]
I like Louie’s argument on debugging and non-IDE editors.

I’ll go through the code…

File: cs/cs/Serializer/VowpalWabbitJsonSerializer.cs (hunk 1 above)
Comment (eisber) [inline cs/cs/Serializer/VowpalWabbitJsonSerializer.cs]: The constants class is already in VW.Serializer. Also the constant is only used in VowpalWabbitJsonSerializer, thus I’d rather keep it there.

File: vowpalwabbit/gd.cc [hunk 4]
```diff
@@ -25,6 +25,7 @@ license as described in the file LICENSE.
 #include "reductions.h"
 #include "vw.h"
 #include "floatbits.h"
+#include <numeric>
```
Comment (JohnLangford) [inline vowpalwabbit/gd.cc]: Can you shift <numeric> to come first in include order?

Comment (JohnLangford) [inline vowpalwabbit/gd.cc]: That is 'numeric'.  Also, why is numeric needed?

**Comment by JohnLangford** [comment]
This says conflicts must be resolved?   

The iterator stuff looks pretty good, but let's try to discuss the json stuff today.

File: vowpalwabbit/gd.cc (hunk 4 above)
Comment (eisber) [inline vowpalwabbit/gd.cc]: I can’t remember why I added it, but the code is gone. I removed it.

**Pull request merged**